R/order.R

```diff
@@ -0,0 +1,101 @@
+#' Order and sort vectors
+#'
+#' @description
+#' `vec_order()` computes the order of `x`. For data frames, the order is
+#' computed along the rows by computing the order of the first column and
+#' using subsequent columns to break ties.
+#'
+#' `vec_sort()` sorts `x` by computing its order and using `vec_slice()` to
+#' rearrange.
+#'
+#' `vec_order_loc()` returns a data frame containing a `key` column with
+#' sorted unique groups, and a `loc` column with the locations of each
+#' group in `x`. It is similar to [vec_group_loc()], except the groups are
+#' returned sorted rather than by first appearance.
+#'
+#' @details
+#' Character vectors are ordered in the C-locale. This is different from
+#' `base::order()`, which respects `base::Sys.setlocale()`, but should
+#' produce more reproducible results between different sessions and platforms.
+#' One immediate difference is that ordering is case-sensitive in the C-locale.
+#' Sorting `c("b", "C", "a")` with `vec_sort()` will return `c("C", "a", "b")`,
+#' but with `base::order()` will return `c("a", "b", "C")` unless
+#' `base::order(method = "radix")` is explicitly set, which also uses the
+#' C-locale. That said, this typically does not introduce any practical issues.
+#'
+#' Character vectors are always sorted in UTF-8. If any non-UTF-8
+#' characters are detected, they are translated to UTF-8 first.
+#'
+#' @param x A vector
+#' @param direction Direction to sort in.
+#'   - Unless `x` is a data frame, this should be a single `"asc"` or
+#'     `"desc"` for ascending or descending order respectively.
+#'   - For data frames, this is allowed to be a vector of `"asc"` and `"desc"`
+#'     with length equal to the number of columns in `x` specifying the
+#'     direction to order each column in.
+#' @param na_value Treatment of `NA` values. `NaN` values are treated as
+#'   equivalent to `NA` values.
+#'   - Unless `x` is a data frame, this should be a single `"largest"` or
+#'     `"smallest"` for treating `NA` values as the largest or smallest values
+#'     respectively.
+#'   - For data frames, this is allowed to be a vector of `"largest"` and
+#'     `"smallest"` with length equal to the number of columns in `x` specifying
+#'     how `NA`s should be treated in each column.
+#' @return
+#' * `vec_order()` an integer vector the same size as `x`.
+#' * `vec_sort()` a vector with the same size and type as `x`.
+#' * `vec_order_loc()`: A two column data frame with size equal to
```

Comment:
Might be worth documenting separately since it's a bit different? And should probably be marked as experimental.

Reply:
It's possible that instead of a separate function this should `vec_group_loc(x, sort = TRUE)`

Reply:
We'd have to think about whether or not it is important to pass along the `direction` and `na_value` arguments somehow vs using the defaults.

Side note: it is also possible to reconstruct `vec_group_loc()` from the internal grouping results of `vec_order()`. You locate the first location of each ordered group, sort those integer locations, and you end up with the ordering of unique keys by first appearance. So it is also worth thinking about if that could be faster in some cases than the current implementation of `vec_group_loc()`